### PR TITLE
System builder

### LINKED ISSUE #79: Refactor World's interface

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This PR introduces a simplified way to create Systems, simplifying the `World` trait (closes #79)

### CODE REVIEW

File: core/src/main/scala/dev/atedeg/ecscala/SystemBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,120 @@
+package dev.atedeg.ecscala
+
+import dev.atedeg.ecscala.util.types.given
+import dev.atedeg.ecscala.util.types.CListTag
+
+trait SystemBuilderOps[L <: CList: CListTag] {
```
Comment (nicolasfara) [inline core/src/main/scala/dev/atedeg/ecscala/SystemBuilder.scala:6]: Missing doc

File: core/src/main/scala/dev/atedeg/ecscala/SystemBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  type BuilderType <: SystemBuilderOps[L]
+  type SystemType <: System[L]
+  def withBefore(before: (DeltaTime, World, View[L]) => Unit): BuilderType
+  def withAfter(after: (DeltaTime, World, View[L]) => Unit): BuilderType
+  def withPrecondition(precondition: => Boolean): BuilderType
+  def withUpdate(f: (Entity, L, DeltaTime) => Deletable[L]): SystemType
+  def withUpdate(f: (Entity, L, DeltaTime, World, View[L]) => Deletable[L]): SystemType
+}
+
+trait SystemBuilder[L <: CList: CListTag] extends SystemBuilderOps[L] {
```
Comment (nicolasfara) [inline core/src/main/scala/dev/atedeg/ecscala/SystemBuilder.scala:16]: Missing doc

File: core/src/main/scala/dev/atedeg/ecscala/SystemBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  override type BuilderType = SystemBuilder[L]
+  override type SystemType = System[L]
+  def excluding[E <: CList: CListTag]: ExcludingSystemBuilder[L, E]
+}
+
+object SystemBuilder {
+  def apply[L <: CList: CListTag]: SystemBuilder[L] = BuilderUtils.SystemBuilderImpl()
+}
+
+trait ExcludingSystemBuilder[L <: CList: CListTag, E <: CList: CListTag] extends SystemBuilderOps[L] {
```
Comment (nicolasfara) [inline core/src/main/scala/dev/atedeg/ecscala/SystemBuilder.scala:26]: Missing doc

**Review by nicolasfara — CHANGES_REQUESTED** [review]

File: core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import org.scalatest.wordspec.AnyWordSpec
+import dev.atedeg.ecscala.util.types.given
+import dev.atedeg.ecscala.fixtures.SystemBuilderFixture
+import dev.atedeg.ecscala.fixtures.{ Position, Velocity }
+
+class SystemBuilderTest extends AnyWordSpec with Matchers {
+  def using = afterWord("using")
+
+  "A SystemBuilder" should {
+    "work with short update" in new SystemBuilderFixture {
```
Comment (vitlinda) [inline core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala:13]: Maybe "work with the minimum number of arguments" or sth like that is better

File: core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      var updateExecuted = false
+      world.addSystem(SystemBuilder[Position &: CNil].withUpdate { (_, c, _) =>
+        updateExecuted = true; c
+      })
+      world.update(10)
+      updateExecuted shouldBe true
+    }
+    "work with full update" in new SystemBuilderFixture {
```
Comment (vitlinda) [inline core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala:21]: Same here: explain the test purpose better

File: core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            updateExecuted = true; c
+          },
+      )
+      world.update(10)
+      updateExecuted shouldBe false
+    }
+  }
+
+  "An ExcludingSystemBuilder" should {
+    "work with short update" in new SystemBuilderFixture {
```
Comment (vitlinda) [inline core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala:108]: Same here

File: core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+      var updateExecuted = false
+      world.addSystem(ExcludingSystemBuilder[Position &: CNil, Velocity &: CNil].withUpdate { (_, c, _) =>
+        updateExecuted = true; c
+      })
+      world.update(10)
+      updateExecuted shouldBe false
+    }
+    "work with full update" in new SystemBuilderFixture {
```
Comment (vitlinda) [inline core/src/test/scala/dev/atedeg/ecscala/SystemBuilderTest.scala:116]: And here

**Review by vitlinda — CHANGES_REQUESTED** [review]

**Review by nicolasfara — APPROVED** [review]

**Review by vitlinda — APPROVED** [review]

**Review by ndido98 — APPROVED** [review]

**Pull request merged**